Support RBAC for cloud-controller-manager for local-up-cluster

dims opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

CCM has a couple of ClusterRoleBinding/ClusterRole(s) that it needs
for cloud-node-controller, pvl-controller and shared-informers service
accounts.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59943

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

dims commented:
/sig auth

dims commented:
/area cloud-provider

dims commented:
Also see earlier somewhat related discussion in https://github.com/kubernetes/kubernetes/pull/56709 about `aws-cloud-provider`

dims commented:
/test pull-kubernetes-node-e2e

dims commented:
/assign @liggitt

dims commented:
`pvl-controller` is from here: 
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L218

references to `shared-informers` are in two locations:
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L204
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L405

references to `cloud-node-controller` are in two locations:
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L210
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L90

liggitt commented:
rootClientBuilder (L204) is not creating a service account named "shared-informers", it just adds a user agent to the requests, and makes them using the credentials the cloud controller manager was started with. I'd expect the same to be done at L405.

The general disposition of things inside cloud-controller-manager is unclear to me. I thought the point of splitting it into a separate process was that it would be extended by things out of tree and run separately. As such, I wasn't expecting the apiserver to bootstrap RBAC permissions for it.

dims commented:
ack @liggitt i'll remove the reference to `shared-informers` and check again

dims commented:
@liggitt guess i feel CCM is special, since we are telling people that they can use that while folks are transitioning to separate repos / binaries / daemonsets.

/sig cluster-lifecycle
/area cloudprovider

liggitt commented:
> @liggitt guess i feel CCM is special, since we are telling people that they can use that while folks are transitioning to separate repos / binaries / daemonsets.

don't we want to start modeling what it looks like to run separately? I thought that would look like a manifest that would deploy the CCM component and set up the required permissions.

dims commented:
/hold

i'll submit another variation

andrewsykim commented:
We had a similar discussion in https://github.com/kubernetes/kubernetes/pull/53511

dims committed:
Support RBAC for cloud-controller-manager for local-up-cluster

CCM has a couple of ClusterRoleBinding/ClusterRole(s) that it needs
for cloud-controller-manager, cloud-node-controller, pvl-controller
and shared-informers service accounts.

dims commented:
/hold cancel

dims commented:
/assign @zmerlynn

dims commented:
/assign @sttts

jhorwit2 reviewed (changes requested):
I think the RBAC files go [here](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/rbac) instead.

jhorwit2 commented:
> The general disposition of things inside cloud-controller-manager is unclear to me. I thought the point of splitting it into a separate process was that it would be extended by things out of tree and run separately. As such, I wasn't expecting the apiserver to bootstrap RBAC permissions for it (since the extensions would likely include different permissions, depending on the cloud)

@liggitt how do you want to handle the service controller which already exists in core since it's part of the KCM currently? Should we bundle it with this even though it's part of the api server bootstrap?

dims commented:
@jhorwit2 Done!

dims commented:
/retest

dims commented:
fyi, https://kubernetes.io/docs/concepts/architecture/cloud-controller/#plugin-mechanism seems to have way lot more. i am running kubetest with local-up-cluster/ccm/openstack and added only those that showed up in the logs.

dims commented:
answering some questions from @cheftako 
```
Hi, sorry your PR is caught in the discussion on the best way to set up the SA/RBAC data. Couple of quick questions. What do we need the CCM in local cluster? What if the api server isn't up when start_cloud_controller_manager runs? I assume the cluster/addons will cause this to be created in any cluster running the static add on-manager. Is there a reasons the local cluster isn't running that add on manager? Do we intend to create the SA/RBAC in clusters which are not (yet) running the CCM?
```

1) need ccm in local cluster for me to run a CI job for testing openstack
2) local-up-cluster.sh has a wait for api server to start and bails out if there is a problem
3) i removed the explicit kubectl for the yaml files from local-up-cluster, will leave it to the addon
4) see #3, so we let something else (or addon) to create the SA/RBAC

dims commented:
@jhorwit2 i added the stuff from https://kubernetes.io/docs/concepts/architecture/cloud-controller/#plugin-mechanism  to my yaml file

liggitt commented:
this LGTM

jhorwit2 commented on cluster/addons/rbac/cloud-controller-manager-roles.yaml:
```diff
@@ -0,0 +1,113 @@
+  metadata:
+    name: cloud-controller-manager
+  rules:
+  - apiGroups:
```
A lot of these permissions are for when we don't use service accounts per controller. The CCM itself needs a subset of these. If i'm not mistaken it just needs get/list/watch for resources (informers), SA creation, events and endpoints.

liggitt replied:
in my opinion, the role should include the permissions you'd need to run the control loop in a pod

jhorwit2 replied:
@liggitt So you're saying the root service account should also have all the permissions of all the controllers?

liggitt replied:
I agree with @jhorwit2... a controller manager should have permissions needed to launch its control loops. that typically means:
* service account setup permissions (read/write service accounts, read access to secrets)
* shared informer permissions (read access to relevant resources)

and that's about it

cheftako replied:
Interestingly I found an orthogonal set of permissions needed to successfully start the CCM. I think rather than having a static list/yaml files in cluster/addons, we would be better of having a list/yaml files in a CP directory and the CP start script should package/copy the list/yaml files into cluster/addons. This also seems like it will more naturally work toward our goal state of allowing each CP to control things like what controllers they run and how many CMs are used to run those controller. It also means the CPs can sign in to having extra permissions as they get things like CM online. So for instance as most CPs don't yet have a CCM running they won't need to add these permissions yet. When they do add them they can better control the permissions given out.

liggitt commented on cluster/addons/rbac/cloud-controller-manager-role-bindings.yaml:
```diff
@@ -0,0 +1,40 @@
+    name: system:cloud-controller-manager
+  subjects:
+  - kind: ServiceAccount
+    name: cloud-controller-manager
```
what runs the cloud controller manager as this service account?

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/59945)
<!-- Reviewable:end -->

dims commented:
/close

moved it to the openstack cloud provider repo
